inst/shiny/server.R

```diff
@@ -218,6 +244,58 @@ shinyServer(function(input, output, session) {
       updateAllPdataInputs()
     }
   })
+ 
+  #Create UI for filtering samples based on annotations
+  output$FilterSamples <- renderUI({
+    annot = ''
+    if (!is.null(vals$counts)){
+      annot = colnames(pData(vals$counts))
+    }
+    selectInput("filteredSample", "Select Annotation:", c("none", annot))
+  })
+  observeEvent(input$filteredSample, {
+    output$filterSampleOptions <- renderUI({
+      if (input$filteredSample != "none")({
+        L = vector("list", 2)
+        L[[1]] = list(checkboxGroupInput("filterSampleChoices", 
+                                         label = "Select samples to keep",
+                                         choices = unique(pData(vals$counts)[,input$filteredSample])))
+        L[[2]] = list(actionButton("runFilterSample", "Filter"))
+        return(L)
+      }) else {
+        L = list()
+      }
+    })
+  })
+  
+  #Filter the selected samples
+  observeEvent(input$runFilterSample, {
+    filter = pData(vals$counts)[,input$filteredSample] %in% input$filterSampleChoices
+    vals$counts = vals$counts[,filter]
+  })
+   
+  #Create UI for filtering genes based on feature annotations
+  observeEvent(input$filteredFeature, {
+    output$filterFeatureOptions <- renderUI({
+      if (input$filteredFeature != "none")({
+        L = vector("list", 2)
+        L[[1]] = list(checkboxGroupInput("filterFeatureChoices", 
+                                         label = "Select features to keep",
+                                         choices = unique(fData(vals$counts)[,input$filteredFeature])))
```

Comment:
@stc1 Couldn't this be potentially 20,000 things long, if a user picks gene name or something like it? Can we avoid rendering this if it has, say 100 or more choices and just send the user a message to pick another column?